packages/chart-advisor/src/advisor/index.ts

```diff
@@ -1,156 +1,113 @@
+   * @param ruleCfg rule base configuration: `include`, `exclude`, `custom`, `options`.
+   *  if not specified, used `ruler` as default
+   */
   constructor(ckbCfg?: CKBConfig, ruleCfg?: RuleConfig) {
     let ckb: Record<string, ChartKnowledgeJSON> = {};
     if (ckbCfg) {
       ckb = this.processCKBCfg(ckbCfg);
     }
 
     this.ckb = ckbCfg ? ckb : CKBJson('en-US', true);
-    this.ruleBase = this.processRuleCfg(ruleCfg);
-
+    this.ruleBase = processRuleCfg(ruleCfg);
   }
 
   /**
    * chart advising from input data
-   * @param input data, DataFrame or raw data
+   * @param data input data as [ { col1: ..., col2, ...}, ... ]
    * @param fields fields to focus
    * @param options advice options such as purpose
    * @returns advice list
    */
-  advise(input: any | DataFrame, fields?: string[], options?: any) {
-    if (input instanceof DataFrame) {
-      this.dataFrame = input;
-    } else {
-      try {
-        this.dataFrame = new DataFrame(input, fields);
-      } catch (error) {
-        // if the input data cannot be transformed into DataFrame
-        console.error('error: ', error);
-        return [];
-      }
+  advise(data: Record<string, any>[], fields?: string[], options?: any) {
+    // transform data into DataFrame
+    let dataFrame: DataFrame;
+    try {
+      dataFrame = new DataFrame(data, { columns: fields });
+    } catch (error) {
+      // if the input data cannot be transformed into DataFrame
+      console.error('error: ', error);
+      return [];
     }
 
-    const advices = dataFrameToAdvices(this.ckb, this.ruleBase, this.dataFrame, options);
+    // get dataProps from dataframe
+    const dataProps = dataFrame.info();
+    const advices = dataToAdvices(data, dataProps as BasicDataPropertyForAdvice[], this.ckb, this.ruleBase, options);
 
     return advices;
   }
 
+  /**
+   * used for testing
+   * @returns ckb in Advisor instance
+   */
   getCKB() {
     return this.ckb;
   }
 
+  /**
+   * used for testing
+   * @returns rule base in the Advisor instance
+   */
   getRuleBase() {
     return this.ruleBase;
   }
 
+  /**
+   * processing ckb config and setup ckb used for advising
+   * @param ckbCfg input ckb configuration
+   * @returns ckb: Record<string, ChartKnowledgeJSON>
+   */
   private processCKBCfg(ckbCfg: CKBConfig) {
-    let ckb: Record<string, ChartKnowledgeJSON> = {};
-
     // step 1: exclude charts from default CKB
-    const defaultCKB = CKBJson('en-US', true);
-    let defaultCKBExcluded: Record<string, ChartKnowledgeJSON> = {};
+    const ckbBase = CKBJson('en-US', true);
     const toExclude = ckbCfg.exclude;
     if (toExclude) {
-      Object.keys(defaultCKB).forEach((chart: string) => {
-        if (!toExclude.includes(chart)) {
-          // charts not being excluded
-          defaultCKBExcluded[chart] = defaultCKB[chart];
+      toExclude.forEach((chartType: string) => {
+        if (Object.keys(ckbBase).includes(chartType)) {
+          delete ckbBase[chartType];
```

Comment:
Use destructruing instead of delete as far as possible.

Reply:
oh i see, I'll update them.

Reply:
fixed